// @fuel-ts/versions/dist/bin.js
#!/usr/bin/env node
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));

// src/cli.ts
var import_chalk2 = require("chalk");
var import_cli_table = __toESM(require("cli-table"));

// src/lib/colorizeUserVersion.ts
var import_chalk = require("chalk");
var colorizeUserVersion = (params) => {
  const { version, isGt, isOk } = params;
  if (isGt) {
    return (0, import_chalk.cyan)(version);
  }
  if (isOk) {
    return (0, import_chalk.green)(version);
  }
  return (0, import_chalk.red)(version);
};

// src/lib/getBuiltinVersions.ts
function getBuiltinVersions() {
  return {
    FORC: "0.49.3",
    FUEL_CORE: "0.22.1",
    FUELS: "0.79.0"
  };
}

// src/lib/semver.ts
function parseVersion(version) {
  const [major, minor, patch] = version.split(".").map((v) => parseInt(v, 10));
  return { major, minor, patch };
}
function versionDiffs(version1, version2) {
  const semver1 = parseVersion(version1);
  const semver2 = parseVersion(version2);
  const major = semver1.major - semver2.major;
  const minor = semver1.minor - semver2.minor;
  const patch = semver1.patch - semver2.patch;
  return {
    major,
    minor,
    patch,
    fullVersionDiff: major || minor || patch
  };
}
function gt(version1, version2) {
  const { fullVersionDiff } = versionDiffs(version1, version2);
  return fullVersionDiff > 0;
}
function eq(version1, version2) {
  const { fullVersionDiff } = versionDiffs(version1, version2);
  return fullVersionDiff === 0;
}

// src/lib/compareSystemVersions.ts
function compareSystemVersions(params) {
  const { systemForcVersion, systemFuelCoreVersion } = params;
  const versions = getBuiltinVersions();
  const systemForcIsGt = gt(systemForcVersion, versions.FORC);
  const systemFuelCoreIsGt = gt(systemFuelCoreVersion, versions.FUEL_CORE);
  const systemForcIsEq = eq(systemForcVersion, versions.FORC);
  const systemFuelCoreIsEq = eq(systemFuelCoreVersion, versions.FUEL_CORE);
  return {
    systemForcIsGt,
    systemFuelCoreIsGt,
    systemForcIsEq,
    systemFuelCoreIsEq
  };
}

// src/lib/fuelUpLink.ts
var fuelUpLink = "https://github.com/fuellabs/fuelup";

// src/lib/getSystemVersions.ts
var import_child_process = require("child_process");
var versionReg = /[0-9]+\.[0-9]+\.[0-9]/;
var getSystemVersion = (command) => {
  let version = null;
  let error = null;
  try {
    const contents = (0, import_child_process.execSync)(command).toString();
    if (versionReg.test(contents)) {
      version = contents.match(versionReg)?.[0];
    } else {
      throw new Error(contents);
    }
  } catch (err) {
    error = err;
  }
  return {
    error,
    version
  };
};
function getSystemForc() {
  const { error, version: v } = getSystemVersion("forc --version");
  return { error, systemForcVersion: v };
}
function getSystemFuelCore() {
  const { error, version: v } = getSystemVersion("fuel-core --version");
  return { error, systemFuelCoreVersion: v };
}
function getSystemVersions() {
  const { error: errorForc, systemForcVersion } = getSystemForc();
  const { error: errorCore, systemFuelCoreVersion } = getSystemFuelCore();
  const error = errorForc ?? errorCore;
  return {
    error,
    systemForcVersion,
    systemFuelCoreVersion
  };
}

// src/cli.ts
var eitherOr = (val1, val2) => val1 ?? val2;
function runVersions() {
  const { error, info } = console;
  const supportedVersions = getBuiltinVersions();
  const cliTable = new import_cli_table.default({
    head: ["", (0, import_chalk2.bold)("Supported"), (0, import_chalk2.bold)(`Yours / System`)]
  });
  const { error: systemError, systemForcVersion, systemFuelCoreVersion } = getSystemVersions();
  const comparisons = compareSystemVersions({
    systemForcVersion: eitherOr(systemForcVersion, "0"),
    systemFuelCoreVersion: eitherOr(systemFuelCoreVersion, "0")
  });
  const userForcColorized = colorizeUserVersion({
    version: eitherOr(systemForcVersion, "\u2014"),
    isGt: comparisons.systemForcIsGt,
    isOk: comparisons.systemForcIsEq
  });
  const userFuelCoreColorized = colorizeUserVersion({
    version: eitherOr(systemFuelCoreVersion, "\u2014"),
    isGt: comparisons.systemFuelCoreIsGt,
    isOk: comparisons.systemFuelCoreIsEq
  });
  cliTable.push(["Forc", supportedVersions.FORC, userForcColorized]);
  cliTable.push(["Fuel-Core", supportedVersions.FUEL_CORE, userFuelCoreColorized]);
  const someIsGt = comparisons.systemForcIsGt || comparisons.systemFuelCoreIsGt;
  const bothAreExact = comparisons.systemForcIsEq && comparisons.systemFuelCoreIsEq;
  let exitCode;
  if (someIsGt) {
    exitCode = 0;
    info(cliTable.toString());
    info(`
Your system's components are newer than the ones supported!`);
  } else if (bothAreExact) {
    exitCode = 0;
    info(cliTable.toString());
    info(`
You have all the right versions! \u26A1`);
  } else if (systemError) {
    exitCode = 1;
    error(cliTable.toString());
    error("\n - Make sure you have Forc and Fuel-Core installed");
    error("   >> Error: ", systemError.message);
  } else {
    exitCode = 1;
    error(cliTable.toString());
    error(`
 - You're using outdated versions`);
  }
  if (exitCode === 1) {
    error(`  ${(0, import_chalk2.green)(fuelUpLink)}`);
  }
  process.exit(exitCode);
}

// src/bin.ts
runVersions();
